### PR TITLE
SLE-616: Update binding view based on project

### PR DESCRIPTION
**Pull request by thahnen**
When a project is closed (or reopened) or deleted, the SonarLint Bindings view should be updated to not include irrelevant projects which might lead to errors when trying to change the binding of it. This does not include importing a project.

### CODE REVIEW

File: org.sonarlint.eclipse.ui/src/org/sonarlint/eclipse/ui/internal/AbstractProjectChangeListener.java [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+
+import org.eclipse.core.resources.IProject;
+import org.eclipse.core.resources.IResourceChangeEvent;
+import org.eclipse.core.resources.IResourceChangeListener;
+import org.eclipse.core.resources.IWorkspace;
+import org.eclipse.core.resources.ResourcesPlugin;
+import org.eclipse.jdt.annotation.Nullable;
+
+public abstract class AbstractProjectChangeListener implements IResourceChangeListener {
+  protected IWorkspace workspace;
```
Comment (henryju) [inline org.sonarlint.eclipse.ui/src/org/sonarlint/eclipse/ui/internal/AbstractProjectChangeListener.java:30]: Not sure storing a reference on the Workspace is a good idea

File: org.sonarlint.eclipse.ui/src/org/sonarlint/eclipse/ui/internal/AbstractProjectChangeListener.java [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+  
+  protected AbstractProjectChangeListener(IProjectChangeHandler handler) {
+    this.workspace = ResourcesPlugin.getWorkspace();
+    this.handler = handler;
+  }
+  
+  @Override
+  public void resourceChanged(IResourceChangeEvent event) {
+    var rsrc = event.getResource();
+    if (rsrc instanceof IProject) {
```
Comment (henryju) [inline org.sonarlint.eclipse.ui/src/org/sonarlint/eclipse/ui/internal/AbstractProjectChangeListener.java:42]: We should avoid relying on IProject (think about our Cobol friends). Instead, you can try to adapt the resource to an ISonarLint project. If it adapts, then work with this ISonarLintProject later on.

File: org.sonarlint.eclipse.ui/src/org/sonarlint/eclipse/ui/internal/binding/BindingsView.java [hunk 3]
```diff
@@ -128,7 +135,7 @@ public void widgetSelected(SelectionEvent e) {
           commonViewer, commonViewer.getNavigatorContentService());
 
         var menuManager = new MenuManager("#PopupMenu");
-        menuManager.addMenuListener(mgr -> {
+        menuManager.addMenuListener((IMenuManager mgr) -> {
```
Comment (henryju) [inline org.sonarlint.eclipse.ui/src/org/sonarlint/eclipse/ui/internal/binding/BindingsView.java:138]: Is this change required?

File: org.sonarlint.eclipse.ui/src/org/sonarlint/eclipse/ui/internal/binding/BindingsView.java [hunk 4]
```diff
@@ -243,7 +251,7 @@ public void connectionRemoved(IConnectedEngineFacade server) {
     };
     SonarLintCorePlugin.getServersManager().addServerLifecycleListener(serverResourceListener);
 
-    serverListener = facade -> {
+    serverListener = (IConnectedEngineFacade facade) -> {
```
Comment (henryju) [inline org.sonarlint.eclipse.ui/src/org/sonarlint/eclipse/ui/internal/binding/BindingsView.java:254]: Is this change required?

File: org.sonarlint.eclipse.ui/src/org/sonarlint/eclipse/ui/internal/binding/BindingsView.java [hunk 5]
```diff
@@ -252,7 +260,17 @@ public void connectionRemoved(IConnectedEngineFacade server) {
...
+    
+    // add listener for when project is closed / deleted
+    AbstractProjectChangeListener.IProjectChangeHandler handler = (IProject project) -> {
+      refreshConnectionState();
+      for (var server : SonarLintCorePlugin.getServersManager().getServers()) {
+        refreshConnectionContent(server);
+      }
+    };
+    
+    projectCloseDeleteListener = new ProjectCloseDeleteListener(handler);
```
Comment (henryju) [inline org.sonarlint.eclipse.ui/src/org/sonarlint/eclipse/ui/internal/binding/BindingsView.java:272]: I am not totally convinced by the need to have 2 different listeners, while in practice you have the same handler in both cases.
Also this AbstractProjectChangeListener seems a bit over-engineered to me. Why not simply having `BindingViews`  implementing IResourceChangeListener? I think this si a common practice in Eclipse codebase.

**Review by henryju — CHANGES_REQUESTED** [review]

File: org.sonarlint.eclipse.ui/src/org/sonarlint/eclipse/ui/internal/binding/BindingsView.java (hunk 3 above)
Comment (thahnen) [reply org.sonarlint.eclipse.ui/src/org/sonarlint/eclipse/ui/internal/binding/BindingsView.java]: No, done

File: org.sonarlint.eclipse.ui/src/org/sonarlint/eclipse/ui/internal/AbstractProjectChangeListener.java (hunk 2 above)
Comment (thahnen) [reply org.sonarlint.eclipse.ui/src/org/sonarlint/eclipse/ui/internal/AbstractProjectChangeListener.java]: I removed the project entirely as it does not work with `ISonarLintProject`

File: org.sonarlint.eclipse.ui/src/org/sonarlint/eclipse/ui/internal/AbstractProjectChangeListener.java (hunk 1 above)
Comment (thahnen) [reply org.sonarlint.eclipse.ui/src/org/sonarlint/eclipse/ui/internal/AbstractProjectChangeListener.java]: Gone, done

File: org.sonarlint.eclipse.ui/src/org/sonarlint/eclipse/ui/internal/binding/BindingsView.java [hunk 6]
```diff
@@ -252,7 +260,42 @@ public void connectionRemoved(IConnectedEngineFacade server) {
     for (var server : SonarLintCorePlugin.getServersManager().getServers()) {
       server.addConnectedEngineListener(serverListener);
     }
-
+    
+    // add listener for when project is opened / closed / deleted
+    projectListener = event -> {
+      if (event.getBuildKind() == IResourceChangeEvent.PRE_CLOSE
+        || event.getBuildKind() == IResourceChangeEvent.PRE_DELETE) {
+        refreshView();
```
Comment (henryju) [inline org.sonarlint.eclipse.ui/src/org/sonarlint/eclipse/ui/internal/binding/BindingsView.java:268]: Won't it fire every time a resource is deleted, like a file or a folder?

File: org.sonarlint.eclipse.ui/src/org/sonarlint/eclipse/ui/internal/binding/BindingsView.java [hunk 7]
```diff
@@ -252,7 +260,42 @@ public void connectionRemoved(IConnectedEngineFacade server) {
...
+      }
+      
+      try {
+        event.getDelta().accept((IResourceDelta delta) -> {
+          var rsrc = delta.getResource();
+          if (((rsrc.getType() & IResource.PROJECT) != 0) 
+            && rsrc.getProject().isOpen()
+            && delta.getKind() == IResourceDelta.CHANGED
+            && ((delta.getFlags() & IResourceDelta.OPEN) != 0)) {
+            refreshView();
```
Comment (henryju) [inline org.sonarlint.eclipse.ui/src/org/sonarlint/eclipse/ui/internal/binding/BindingsView.java:283]: Shouldn't you return "false" after that, to prevent multiple calls to "refresh". In fact, it might be more robust to collect a boolean "shouldRefresh", and only call "refreshView" outside the delta visitor.

**Review by henryju — CHANGES_REQUESTED** [review]

File: org.sonarlint.eclipse.ui/src/org/sonarlint/eclipse/ui/internal/binding/BindingsView.java (hunk 6 above)
Comment (thahnen) [reply org.sonarlint.eclipse.ui/src/org/sonarlint/eclipse/ui/internal/binding/BindingsView.java]: No the event is only for projects

File: org.sonarlint.eclipse.ui/src/org/sonarlint/eclipse/ui/internal/binding/BindingsView.java (hunk 7 above)
Comment (thahnen) [reply org.sonarlint.eclipse.ui/src/org/sonarlint/eclipse/ui/internal/binding/BindingsView.java]: Done

File: org.sonarlint.eclipse.ui/src/org/sonarlint/eclipse/ui/internal/binding/BindingsView.java [hunk 8]
```diff
@@ -252,7 +262,51 @@ public void connectionRemoved(IConnectedEngineFacade server) {
...
+        || event.getType() == IResourceChangeEvent.PRE_DELETE) {
+        shouldRefresh = true;
+      } else if (event.getDelta() != null) {
+        try {
+          event.getDelta().accept((IResourceDelta delta) -> {
+            if (delta.getKind() == IResourceDelta.CHANGED) {
+              // INFO: With adding "IResourceDelta.ADDED" to the mask we can also update the binding view when
+              //       importing a project. Therefore the user can see if a newly imported project is already
+              //       correctly binded.
+              for (var child : delta.getAffectedChildren(IResourceDelta.CHANGED)) {
```
Comment (thahnen) [inline org.sonarlint.eclipse.ui/src/org/sonarlint/eclipse/ui/internal/binding/BindingsView.java:278]: FYI: The delta resource is the Workspace and therefore we have to check if the direct children have changed. We specifically only look for changed projects (excluding hidden ones like "External Plug-in Libraries").

**Comment by thahnen** [comment]
Adapting to ISonarLintProject did not work at all, tested that multiple times on multiple levels. If this very feature does not work on Cobol editors, I have to say it is a risk I'm willing to take. This way it can be a QOL improvement at least for the other developers.

File: org.sonarlint.eclipse.ui/src/org/sonarlint/eclipse/ui/internal/binding/BindingsView.java [hunk 9]
```diff
@@ -252,7 +262,51 @@ public void connectionRemoved(IConnectedEngineFacade server) {
     for (var server : SonarLintCorePlugin.getServersManager().getServers()) {
       server.addConnectedEngineListener(serverListener);
     }
-
+    
+    // add listener for when project is opened / closed / deleted
+    projectListener = event -> {
+      if (event.getType() == IResourceChangeEvent.PRE_CLOSE
```
Comment (henryju) [inline org.sonarlint.eclipse.ui/src/org/sonarlint/eclipse/ui/internal/binding/BindingsView.java:268]: Why not declaring the `shouldRefresh` variable here, instead of risking a race condition?

**Review by henryju — APPROVED** [review]
I would move the `shouldRefresh` field to a local variable, then LGTM

File: org.sonarlint.eclipse.ui/src/org/sonarlint/eclipse/ui/internal/binding/BindingsView.java (hunk 9 above)
Comment (thahnen) [reply org.sonarlint.eclipse.ui/src/org/sonarlint/eclipse/ui/internal/binding/BindingsView.java:268]: Cannot be done 😞 When used as well in `IResourceDelta.accept(...)` in line 273+ it cannot be compiled with the issue: "Variable shouldRefresh used in lambda expression should be final or effectively final".
The ideal resolution would be to use a class variable and in case of the listener being delegated by the workspace to different threads (which I cannot foresee) to use `volatile`.

@henryju regarding a race condition: From manual testing, it seems that the events are processed via a queue but I'm not sure if it is really the case. I can move it to an `AtomicBoolean` maybe?

Comment (henryju) [reply org.sonarlint.eclipse.ui/src/org/sonarlint/eclipse/ui/internal/binding/BindingsView.java:268]: Yeah, I think a local AtomicBoolean is better in my opinion

Comment (thahnen) [reply org.sonarlint.eclipse.ui/src/org/sonarlint/eclipse/ui/internal/binding/BindingsView.java:269]: Done
